tests/conftest.py

```diff
@@ -0,0 +1,132 @@
+
+    os.environ['USE_MIDDLEWARE'] = "False"
+    relay_port = find_free_port()
+    os.environ['NETAPP_PORT'] = str(relay_port)
+    os.environ["NETAPP_ADDRESS"] = f"http://0.0.0.0:{relay_port}"
+
+
+@pytest.fixture(scope="class")
+def run_relay_server(set_env_common):
+    """ Run Relay Network Application Interface"""
+
+    server_env = os.environ.copy()
+    server_env["ROS_DOMAIN_ID"] = os.environ["ROS_DOMAIN_ID_SERVER"]
+    
+    if USING_PANTS:
+        proc_cmd = ["era_5g_relay_network_application/server.pex"]
+    else:
+        proc_cmd = ["python3", "-m", "era_5g_relay_network_application.server"]
+
+    print(f"Starting Relay Server, domain id: {server_env['ROS_DOMAIN_ID']}")
+    
+    # Relay Interface can create more processes, so os.setsid() is
+    # used to create a group, and as a result subsequent term signal 
+    # is later sent to all of the child processes of the group
+    proc = sp.Popen(proc_cmd, preexec_fn=os.setsid, env=server_env, stdout=sp.PIPE, stderr=sp.STDOUT) 
+
+    pid = proc.pid
+    
+    if proc.poll() is not None:
+        print(f"Problem occured when starting Relay Server: {proc.communicate()}")
+
+    yield
+    
+    # Terminate process and all other processes it created
+    os.killpg(os.getpgid(pid), signal.SIGKILL)
+
+    # Wait for the process to terminate 
+    # (this makes sure there is no interference between multiple runs)
+    os.waitpid(pid, 0)
+
+
+@pytest.fixture(scope="class")
+def run_relay_client(set_env_common, run_relay_server):
+    """Run Relay Network Application Client"""
+
+    # Wait to let the Relay Server start up
+    time.sleep(2)
+
+    client_env = os.environ.copy()
+    client_env["ROS_DOMAIN_ID"] = os.environ["ROS_DOMAIN_ID_CLIENT"]
+    client_env["WAIT_UNTIL_AVAILABLE"] = "true"
+    client_env["WAIT_TIMEOUT"] = str(10)
+
+    if USING_PANTS:
+        proc_cmd = ["era_5g_relay_network_application/client.pex"]
+    else:
+        proc_cmd = ["python3", "-m", "era_5g_relay_network_application.client"]
+
+    print(f"Starting Relay Client, domain id: {client_env['ROS_DOMAIN_ID']}")
+    proc = sp.Popen(proc_cmd, preexec_fn=os.setsid, env=client_env)
+    pid = proc.pid
+    if proc.poll() is not None:
+        print(f"Problem occured when starting Relay Client: {proc.communicate()}")
+    yield
+    os.killpg(os.getpgid(pid), signal.SIGKILL)
+    os.waitpid(pid, 0)
+
+
+@pytest.fixture(scope="function")
+def client_executor(set_env_common):
+    """Create ROS executor for server"""
+
+    os.environ["ROS_DOMAIN_ID"] = os.environ["ROS_DOMAIN_ID_CLIENT"]
+    rclpy.init()
+    ros_executor = MultiThreadedExecutor()
+    executor_thread = Thread(target=ros_executor.spin, daemon=True)
+    executor_thread.start()
+    yield ros_executor
+    rclpy.shutdown()
+
+
+@pytest.fixture(scope="function")
+def server_executor(set_env_common):
+    """Create ROS executor for client"""
+
+    os.environ["ROS_DOMAIN_ID"] = os.environ["ROS_DOMAIN_ID_SERVER"]
```

Comment:
Does this really work? Because you are altering the global environment...

Reply:
I will check this again to make sure.

Reply:
I double checked that this approach is working fine. When turning off Relay Client and/or Server (simply by not providing the fixture) the test fails. Similar behaviour can also be simulated by not providing the correct variable, e. g. TOPICS_TO_SERVER. 

The approach works because all processes related to a test are run independently and each process has its own env. The main env is also changed here (in server_executor and client_executor), but all other processes are independent.

However, when experimenting with the tests, I found some other things that should be improved, so I chaned them. There are also new changes related to renaming of the env variables.